galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,311 @@
+type Handler struct {
+	domainSuffix string
+
+	listener processing.Listener
+
+	services     map[resource.FullName]resource.Entry
+	endpoints    map[resource.FullName]resource.Entry
+	podsHandler  processing.Handler
+	nodesHandler processing.Handler
+	pods         pod.Cache
+	nodes        node.Cache
+
+	// The version number for the current State of the object. Every time mcpResources or versions change,
+	// the version number also change
+	version      int64
+	mcpLock      sync.Mutex
+	mcpResources map[resource.FullName]*mcp.Resource
+
+	// pendingEvents counts the number of events awaiting publishing.
+	pendingEvents int64
+
+	// lastSnapshotTime records the last time a snapshot was published.
+	lastSnapshotTime time.Time
+}
+
+// NewHandler creates a new Handler instance.
+func NewHandler(domainSuffix string, listener processing.Listener) *Handler {
+	pods, podsHandler := pod.NewCache()
+	nodes, nodesHandler := node.NewCache()
+
+	return &Handler{
+		domainSuffix: domainSuffix,
+		listener:     listener,
+		services:     make(map[resource.FullName]resource.Entry),
+		endpoints:    make(map[resource.FullName]resource.Entry),
+		mcpResources: make(map[resource.FullName]*mcp.Resource),
+		podsHandler:  podsHandler,
+		nodesHandler: nodesHandler,
+		pods:         pods,
+		nodes:        nodes,
+	}
+}
+
+// Handle incoming events and generate synthetic ServiceEntry projections.
+func (p *Handler) Handle(event resource.Event) {
+	switch event.Entry.ID.Collection {
+	case metadata.K8sCoreV1Endpoints.Collection:
+		// Update the projections
+		p.handleEndpointsEvent(event)
+	case metadata.K8sCoreV1Services.Collection:
+		// Update the projections
+		p.handleServiceEvent(event)
+	case metadata.K8sCoreV1Nodes.Collection:
+		// Just add the node to the cache.
+		p.nodesHandler.Handle(event)
+	case metadata.K8sCoreV1Pods.Collection:
+		// Just add the pod to the cache.
+		p.podsHandler.Handle(event)
+	default:
+		scope.Warnf("received event with unexpected collection: %v", event.Entry.ID.Collection)
+	}
+}
+
+// Builds the snapshot of the current resources.
+func (p *Handler) BuildSnapshot() snapshot.Snapshot {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	now := time.Now()
+	monitoring.RecordProcessorSnapshotPublished(p.pendingEvents, now.Sub(p.lastSnapshotTime))
+	p.lastSnapshotTime = now
+	p.pendingEvents = 0
+
+	b := snapshot.NewInMemoryBuilder()
+
+	// Copy the entries.
+	entries := make([]*mcp.Resource, 0, len(p.mcpResources))
+	for _, r := range p.mcpResources {
+		entries = append(entries, r)
+	}
+
+	// Create the collection resources on the Snapshot.
+	version := fmt.Sprintf("%d", p.version)
+	b.Set(collection.String(), version, entries)
+
+	return b.Build()
+}
+
+func (p *Handler) handleServiceEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the service's version has changed.
```

Comment:
When would we see this on Add?

Reply:
I suspect we get an add from k8s when the resource was just created.  Not sure, but we might also get it when the resource is new WRT our  k8s client's local cache.